src/sentry_core.c

```diff
@@ -89,39 +90,61 @@ sentry_init(sentry_options_t *options)
     // enumeration.
     options->run = sentry__run_new(options->database_path);
     if (!options->run) {
-        sentry_options_free(options);
-        return 1;
+        SENTRY_WARN("failed to initialize run directory");
+        goto fail;
     }
 
     load_user_consent(options);
 
-    // we "activate" the options here, since the transport and backend might try
-    // to access them
+    if (transport) {
+        if (!sentry__transport_startup(transport, options)) {
+            SENTRY_WARN("failed to initialize transport");
+            goto fail;
+        }
+    }
+
+    // and then we will start the backend, since it requires a valid run
+    sentry_backend_t *backend = options->backend;
+    if (backend && backend->startup_func) {
+        SENTRY_TRACE("starting backend");
+        if (!backend->startup_func(backend, options)) {
+            SENTRY_WARN("failed to initialize backend");
+            goto fail;
+        }
+    }
+
     sentry__mutex_lock(&g_options_mutex);
     g_options = options;
     sentry__mutex_unlock(&g_options_mutex);
 
-    sentry_transport_t *transport = options->transport;
-    if (transport) {
-        sentry__transport_startup(transport, options);
+    // *after* setting the global options, trigger a scope and consent flush,
+    // since at least crashpad needs that.
+    // the only way to get a reference to the scope is by locking it, the macro
+    // does all that at once, including invoking the backends scope flush hook
+    SENTRY_WITH_SCOPE_MUT (scope) {
+        (void)scope;
+    }
+    if (backend && backend->user_consent_changed_func) {
+        backend->user_consent_changed_func(backend);
     }
 
     // after initializing the transport, we will submit all the unsent envelopes
     // and handle remaining sessions.
     sentry__process_old_runs(options);
 
-    // and then we will start the backend, since it requires a valid run
-    sentry_backend_t *backend = options->backend;
-    if (backend && backend->startup_func) {
-        SENTRY_TRACE("starting backend");
-        backend->startup_func(backend, options);
-    }
-
     if (options->auto_session_tracking) {
         sentry_start_session();
     }
 
     return 0;
+
+fail:
+    SENTRY_WARN("`sentry_init` failed");
+    if (transport) {
+        sentry__transport_shutdown(transport, 0);
+    }
+    sentry_options_free(options);
```

Comment:
Should we also free the backend here if it is set?

Reply:
The backend is freed inside of `options_free`. I thought about doing the shutdown, but the backend start is the last thing that can fail, so doing a shutdown does not make sense.